scripts/benchmark/checks/erc20_interact.py

```diff
@@ -42,41 +44,70 @@
     new_transaction,
 )
 
+from enum import (
+    Enum,
+)
+
 
 EXPECTED_TOTAL_SUPPLY = 10000000000000000000000
 FIRST_TX_GAS_LIMIT = 1400000
-SECOND_TX_GAS_LIMIT = 22000
+SECOND_TX_GAS_LIMIT = 60000
 
 CONTRACT_FILE = 'scripts/benchmark/contract_data/erc20.sol'
 CONTRACT_NAME = 'SimpleToken'
 
 W3_TX_DEFAULTS = {'gas': 0, 'gasPrice': 0}
 
 
-class DeployErc20(BaseBenchmark):
+class ERC20BenchmarkMode(Enum):
```

Comment:
I think this would be better structured if we dropped the `mode` concept and instead used subclasses of base class to implement each benchmark.

```
class BaseERC20Benchmark(BaseBenchmark):
    def execute(self):
        ... # setup code here
        self._do_benchmark(...)

class ERC20TransferBenchmark(BaseERC20Benchmark):
    name = 'ERC20 Transfer Benchmark'

    def _do_benchmark(self):
        ... # logic for the `tranfer` test
```

Reply:
@pipermerriam the one thing to keep in mind here though is that we are currently thinking of running these benchmarks in a defined order, with the benchmark that deploys the contract as the first one, and then snapshot the chain to perform the transfer/approve benchmarks on top of an already deployed contract rather than redeploying the contract again (to have a better isolation of *what* is being tested).

Not saying, this can not be done in the inheritance model but it may be slightly more pragmatic in the one single instance, three different calls model.

No strong opinion, just weighing in here to give some perspective. But I'd say we can shoot for *any* approach that achieves the "only test `transfer` / `approve` without redeploying the contract" criteria and then look at how to maybe refactor that into the architecture we want.

Reply:
@cburgdorf so after making the persistent chain, I'm not sure if it makes the most sense, as it requires the overhead of holding onto the chains through multiple benchmarks in memory, and the difference in benchmark execution time is very very small, (I can get an exact measurement if needed). 

I also don't think that the persistent chain really helps with isolation, because wthe initial setup transactions do not get included in the benchmarked data at all.  The setup transactions are sent and mined before any of the timing, and benchmarking even begins, and is not included in any of the benchmarked data or any of the totals. 

I also think that it makes things a bit harder to understand, but I think that is more my fault then anything, but still something that needs to be addressed.

side note: when I was testing initially, before realized to revert the canonical head of the chain, I found that if I ran the full transfer benchmark twice, the second time it would run more then a second slower in every VM version, because the head wouldn't reset, and the chain would start mining on block 201. (this is fixed by resetting the canonical head of the chain) but just wanted to point that out. Im not sure if that time difference is noteworthy but it seemed so to me

Reply:
>The setup transactions are sent and mined before any of the timing, and benchmarking even begins, and is not included in any of the benchmarked data or any of the totals.

Yes, I think moving the contract deployment into a setup step that is run before the measurement actually starts will also work (as long as the setup time is small enough which is what you suggest). So, I'm ok to use that as an approach especially as it makes the benchmarks less inter dependent.

>side note: when I was testing initially, before realized to revert the canonical head of the chain, I found that if I ran the full transfer benchmark twice, the second time it would run more then a second slower in every VM version, because the head wouldn't reset, and the chain would start mining on block 201

Interesting. There's probably a reason for that. One second slower seems like a big difference here but my guess would be that there is a strong difference between starting on an empty chain vs starting on a chain that already has a couple of hundred blocks but that the difference won't scale linearly onwards from here. Just guessing though.

Reply:
on my machine running only the ERC20 benchmarks blocks = 100, tx/block = 2:

Persistent chain benchtime : 87.543 seconds
Persistent chain total run time : 87.548 seconds

new chain benchtime : 86.435 seconds
new chain total run time : 87.269 seconds